benchmarks/src/test/java/org/apache/druid/benchmark/DruidSchemaInternRowSignatureBenchmark.java

```diff
@@ -0,0 +1,219 @@
+      );
+    }
+
+    @Override
+    public Set<SegmentId> refreshSegments(final Set<SegmentId> segments) throws IOException
+    {
+      return super.refreshSegments(segments);
+    }
+
+
+    @Override
+    public void addSegment(final DruidServerMetadata server, final DataSegment segment)
+    {
+      super.addSegment(server, segment);
+    }
+
+    @Override
+    protected Sequence<SegmentAnalysis> runSegmentMetadataQuery(Iterable<SegmentId> segments)
+    {
+      final int numColumns = 1000;
+      Map<String, ColumnAnalysis> columnToAnalysisMap = new HashMap<>();
+      for (int i = 0; i < numColumns; ++i) {
+        columnToAnalysisMap.put(
+            "col" + i,
+            new ColumnAnalysis(
+                ColumnType.STRING,
+                null,
+                false,
+                false,
+                40,
+                null,
+                null,
+                null,
+                null
+            )
+        );
+      }
+      return Sequences.simple(
+          Lists.transform(
+              Lists.newArrayList(segments),
+              (segment) -> new SegmentAnalysis(
+                  segment.toString(),
+                  ImmutableList.of(segment.getInterval()),
+                  columnToAnalysisMap,
+                  40,
+                  40,
+                  null,
+                  null,
+                  null,
+                  false
+              )
+          )
+      );
+    }
+  }
+
+  @State(Scope.Thread)
+  public static class MyState
+  {
+    SegmentId segmentId;
+
+    @Setup(Level.Iteration)
+    public void setup()
+    {
+      segmentId = SegmentId.of("dummy", Intervals.ETERNITY, "1", new LinearShardSpec(0));
+    }
+
+    @TearDown(Level.Iteration)
+    public void teardown()
+    {
+      segmentId = null;
+    }
+  }
+
+  @Setup
+  public void setup()
+  {
+
+    druidSchema = new DruidSchemaForBenchmark(
+        EasyMock.mock(QueryLifecycleFactory.class),
+        EasyMock.mock(TimelineServerView.class),
+        null,
+        null,
+        EasyMock.mock(PlannerConfig.class),
+        null,
+        null
+    );
+    serverMetadata = new DruidServerMetadata(
+        "dummy",
+        "dummy",
+        "dummy",
+        42,
+        ServerType.HISTORICAL,
+        "tier-0",
+        0
+    );
+
+    dataSegment = DataSegment.builder()
+                             .dataSource("dummy")
+                             .shardSpec(new LinearShardSpec(0))
+                             .dimensions(ImmutableList.of("col1", "col2", "col3", "col4"))
+                             .version("1")
+                             .interval(Intervals.ETERNITY)
+                             .size(0)
+                             .build();
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  public void addSegments(MyState state, Blackhole blackhole) throws IOException
+  {
+    druidSchema.addSegment(serverMetadata, dataSegment);
```

Comment:
I feel we should rather pass a list of like 10K segments than calling the same method with single segment 10K times. Former is a more closer to how it will be called in the real-world.

Reply:
Updated the benchmark to call updates on a large set of segments at once, and reduced the number of iterations that it does this over.